test/util/dynamic/dynamic.go

```diff
@@ -0,0 +1,89 @@
+package dynamic
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+// Client returns a client from a given unstructured object.
+// It can be used when creating an object from a yaml file.
+type Client interface {
+	GetClient(obj *unstructured.Unstructured) (ResourceClient, error)
+}
+
+type client struct {
+	dynamic.Interface
+	mapping meta.RESTMapper
+}
+
+func NewDynamicClient(kubeConfig *rest.Config) (Client, error) {
+	cli, err := dynamic.NewForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	groupResources, err := restmapper.GetAPIGroupResources(discoveryClient)
+	if err != nil {
+		return nil, err
+	}
+	mapper := restmapper.NewDiscoveryRESTMapper(groupResources)
```

Comment:
If at all possible, use a deferred discovery mapper - otherwise, this will eagerly send a request storm to grab all discovery information at once.